engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/Atom.java

```diff
@@ -20,184 +20,179 @@
-   *
-   * @return a textual representation of this Atom.
-   */
-  @Override
-  public String toString() {
-    return toString(10);
-  }
-
-  @CompilerDirectives.TruffleBoundary
-  private String toString(int depth) {
-    StringBuilder sb = new StringBuilder();
-    toString(sb, false, depth);
-    return sb.toString();
-  }
-
-  @ExportMessage
-  public boolean hasMembers() {
-    return true;
-  }
-
-  @ExportMessage
-  @CompilerDirectives.TruffleBoundary
-  public Array getMembers(boolean includeInternal) {
-    Map<String, Function> members = constructor.getDefinitionScope().getMethods().get(constructor.getType());
-    if (members == null) {
-      return new Array(0);
+
+    @CompilerDirectives.TruffleBoundary
+    private String toString(int depth) {
+        StringBuilder sb = new StringBuilder();
+        toString(sb, false, depth);
+        return sb.toString();
     }
-    Object[] mems = members.keySet().toArray();
-    return new Array(mems);
-  }
-
-  @ExportMessage
-  @CompilerDirectives.TruffleBoundary
-  public boolean isMemberInvocable(String member) {
-    Map<String, ?> members = constructor.getDefinitionScope().getMethods().get(constructor.getType());
-    return members != null && members.containsKey(member);
-  }
-
-  @ExportMessage
-  @ExplodeLoop
-  public boolean isMemberReadable(String member) {
-    for (int i = 0; i < constructor.getArity(); i++) {
-      if (member.equals(constructor.getFields()[i].getName())) {
+
+    @ExportMessage
+    public boolean hasMembers() {
         return true;
-      }
     }
-    return false;
-  }
-
-  @ExportMessage
-  @ExplodeLoop
-  public Object readMember(String member) throws UnknownIdentifierException {
-    for (int i = 0; i < constructor.getArity(); i++) {
-      if (member.equals(constructor.getFields()[i].getName())) {
-        return fields[i];
-      }
+
+    @ExportMessage
+    @CompilerDirectives.TruffleBoundary
+    public Array getMembers(boolean includeInternal) {
+        Map<String, Function> members = constructor.getDefinitionScope().getMethods().get(constructor.getType());
+        if (members == null) {
+            return new Array(0);
+        }
+        Object[] mems = members.keySet().toArray();
+        return new Array(mems);
     }
-    throw UnknownIdentifierException.create(member);
-  }
 
-  @ExportMessage
-  static class InvokeMember {
+    @ExportMessage
+    @CompilerDirectives.TruffleBoundary
+    public boolean isMemberInvocable(String member) {
+        Map<String, ?> members = constructor.getDefinitionScope().getMethods().get(constructor.getType());
+        return members != null && members.containsKey(member);
+    }
 
-    static UnresolvedSymbol buildSym(AtomConstructor cons, String name) {
-      return UnresolvedSymbol.build(name, cons.getDefinitionScope());
+    @ExportMessage
+    @ExplodeLoop
+    public boolean isMemberReadable(String member) {
+        for (int i = 0; i < constructor.getArity(); i++) {
```

Comment:
This is not going to `@ExplodeLoop` I am afraid. `constructor` isn't compilation constant. You'd need some profile, I think.